### PR TITLE
[HotFix][CARBONDATA-2788][BloomDataMap] Fix bugs in incorrect query result with bloom datamap

### PR DESCRIPTION
**Pull request by xuchuanyin**
This PR solve two problems which will affect the correctness of the query on bloom.

1. Revert PR2539

After review the code, we found that modification in PR2539 is not needed, so we revert that PR.

2. bugs in overflow for blocklet count

Carbondata stores blocklet count for each block in byte data type, when
a block contains more than 128 blocklets, it will overflow the byte
limits.
Here we change the data type to short.

3. For cache_leve=block, after pruning by main BlockDataMap, the blockletNo
in Blocklet is -1, which indicate that the following percedure will scan
the whole block -- *all* the blocklets in the block.
So, when doing intersection with the pruned result from BloomDataMap, we need to take care of these blocklets. In this implementation, we added the result from BloomDataMap based on the blocklet's existence in BlockDataMap.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7538/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6292/

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java [hunk 1]
```diff
@@ -103,7 +106,19 @@ public void init(DataMapModel dataMapModel) throws IOException {
...
+  public void initIndexColumnConverters(CarbonTable carbonTable, String dataMapName,
+      List<CarbonColumn> indexedColumn) {
+    String cacheLevel = MapUtils.getString(
+        carbonTable.getTableInfo().getFactTable().getTableProperties(),
+        CarbonCommonConstants.CACHE_LEVEL, CarbonCommonConstants.CACHE_LEVEL_DEFAULT_VALUE);
+    this.isBlockletCacheLevel = cacheLevel.equalsIgnoreCase("blocklet");
+    if (!this.isBlockletCacheLevel) {
+      LOGGER.warn(
+          String.format("BloomFilter datamap %s runs with cache_level=block for table %s.%s,"
+              + " which may decrease its pruning performance",
```
Comment (jackylk) [inline datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java]: change to `which may decrease its pruning benefit, which lead to read more data`

File: integration/spark2/src/main/scala/org/apache/carbondata/datamap/IndexDataMapRebuildRDD.scala [hunk 2]
```diff
@@ -357,13 +357,20 @@ class IndexDataMapRebuildRDD[K, V](
         // skip clear datamap and we will do this adter rebuild
         reader.setSkipClearDataMapAtClose(true)
 
+        // currently blockletId in rowWithPosition is wrong, we cannot use it
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/carbondata/datamap/IndexDataMapRebuildRDD.scala]: This is a bit confusing, can you rephrase it

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java (hunk 1 above)
Comment (xuchuanyin) [reply datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java]: OK

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7543/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6297/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7545/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6299/

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java [hunk 3]
```diff
@@ -165,9 +180,11 @@ public void initIndexColumnConverters(CarbonTable carbonTable, List<CarbonColumn
       for (CarbonBloomFilter bloomFilter : bloomIndexList) {
         boolean scanRequired = bloomFilter.membershipTest(new Key(bloomQueryModel.filterValue));
         if (scanRequired) {
+          String blockletNo =
+              isBlockletCacheLevel ? String.valueOf(bloomFilter.getBlockletNo()) : "-1";
```
Comment (manishgupta88) [inline datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java]: I think Bloom dataMap should return the actual blocklet Id. The old behavior should not be modified based on cache level

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6020/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7567/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6322/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7573/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7575/

**Comment by xuchuanyin** [comment]
Till now, there is still a problem: for the test case added in Bloom*FunctionSuite, the explain query output will give negative pruned result like:
```
|== CarbonData Profiler ==
Table Scan on test_rcd
 - total blocklets: 1
 - filter: (city <> null and city = city40)
 - pruned by Main DataMap
    - skipped blocklets: 0
 - pruned by CG DataMap
    - name: dm_rcd
    - provider: bloomfilter
    - skipped blocklets: -1
                                    |
|== Physical Plan ==
*FileScan carbondata default.test_rcd[id#172,country#173,city#174,population#175,random1#176,random2#177,random3#178,random4#179,random5#180,random6#181,random7#182,random8#183,random9#184,random10#185,random11#186,random12#187] PushedFilters: [IsNotNull(city), EqualTo(city,city40)]|
```

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7577/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6327/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6332/

**Comment by xuchuanyin** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7579/

**Comment by chenliang613** [comment]
retest this please

**Comment by chenliang613** [comment]
i tested at my machine ,  the PR is working fine for query data with bloomfilter.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7580/

File: datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java (hunk 3 above)
Comment (xuchuanyin) [reply datamap/bloom/src/main/java/org/apache/carbondata/datamap/bloom/BloomCoarseGrainDataMap.java]: ok, fixed

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7581/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6336/

**Comment by xuchuanyin** [comment]
Hi, all @jackylk @ravipesala @manishgupta88 @chenliang613 
I raised another PR #2574 as another implementation for this PR, please check it also

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6039/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6041/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6042/

**Comment by xuchuanyin** [comment]
PR2574 is to replace this PR

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7590/

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonInputFormat.java [hunk 4]
```diff
@@ -492,6 +493,26 @@ protected Expression getFilterPredicates(Configuration configuration) {
...
+      List<ExtendedBlocklet> previousDataMapPrunedBlocklets,
+      List<ExtendedBlocklet> otherDataMapPrunedBlocklets) {
+    List<ExtendedBlocklet> prunedBlocklets = null;
+    if (BlockletDataMapUtil
+        .isCacheLevelBlock(carbonTable, BlockletDataMapFactory.CACHE_LEVEL_BLOCKLET)) {
+      prunedBlocklets = new ArrayList<>(otherDataMapPrunedBlocklets);
+      // add blocklets from previous dataMap that are not filtered by other dataMaps
+      for (ExtendedBlocklet previousBlocklet : previousDataMapPrunedBlocklets) {
+        if (!otherDataMapPrunedBlocklets.contains(previousBlocklet)) {
+          prunedBlocklets.add(previousBlocklet);
```
Comment (ravipesala) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonInputFormat.java]: It supposed to be same as with blocklet level. why adding extra non pruned blocklets should be added to the list? Any use case for this check?

Comment (xuchuanyin) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonInputFormat.java]: Fixed as we talked

**Comment by xuchuanyin** [comment]
As for the incorrect explain output, I raised a jira 2797 to track this.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7602/

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapFactory.java [hunk 5]
```diff
@@ -71,7 +71,7 @@
   /**
    * variable for cache level BLOCKLET
    */
-  private static final String CACHE_LEVEL_BLOCKLET = "BLOCKLET";
+  public static final String CACHE_LEVEL_BLOCKLET = "BLOCKLET";
```
Comment (chenliang613) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapFactory.java]: why changes to public

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java [hunk 6]
```diff
@@ -318,13 +318,22 @@ private DataMapRowImpl loadBlockMetaInfo(CarbonRowSchema[] taskSummarySchema,
...
     }
-    byte[] blockletCount = ArrayUtils
-        .toPrimitive(blockletCountInEachBlock.toArray(new Byte[blockletCountInEachBlock.size()]));
+    byte[] blockletCount = convertRowCountFromShortToByteArray(blockletCountInEachBlock);
     // blocklet count index is the last index
     summaryRow.setByteArray(blockletCount, taskSummarySchema.length - 1);
     return summaryRow;
   }
 
+  private byte[] convertRowCountFromShortToByteArray(List<Short> blockletCountInEachBlock) {
```
Comment (chenliang613) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java]: why need to do the convert ?

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/datamap/lucene/LuceneFineGrainDataMapSuite.scala [hunk 7]
```diff
@@ -371,7 +371,7 @@ class LuceneFineGrainDataMapSuite extends QueryTest with BeforeAndAfterAll {
       """
         | CREATE TABLE datamap_test_table(id INT, name STRING, city STRING, age INT)
         | STORED BY 'carbondata'
-        | TBLPROPERTIES('SORT_COLUMNS'='city,name', 'SORT_SCOPE'='GLOBAL_SORT')
+        | TBLPROPERTIES('SORT_COLUMNS'='city,name', 'SORT_SCOPE'='GLOBAL_SORT', 'CACHE_LEVEL'='BLOCKLET')
```
Comment (chenliang613) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/datamap/lucene/LuceneFineGrainDataMapSuite.scala]: why need change "CACHE_LEVEL" to "BLOCKLET"

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7640/

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapFactory.java (hunk 5 above)
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapFactory.java]: Because this member needs to be accessed outside this class. Currently in `CarbonInputFormat` we need to use this variable to know the current cache level.

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java (hunk 6 above)
Comment (xuchuanyin) [reply core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java]: because we are using offheap store, which needs to  store the bytes.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/datamap/lucene/LuceneFineGrainDataMapSuite.scala (hunk 7 above)
Comment (xuchuanyin) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/datamap/lucene/LuceneFineGrainDataMapSuite.scala]: By default the cache_level is BLOCK which may affect the pruning info. In some test cases in this file, they assert on the content of pruning info. So here, I just change the cache_level to BLOCKLET, so that I do not to modify the assertion.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6365/

**Comment by manishgupta88** [comment]
LGTM

**Comment by jackylk** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6382/

**Comment by chenliang613** [comment]
verified, LGTM
